### PR TITLE
Rewrite xml parser

### LINKED ISSUE #53: Rewrite tool parser
The current solution based in SAX is a bit Ad-Hoc and somewhat hacky, especially when dealing with incomplete or malformed XML.
I have experimented with other existing pythonic XML parsers, like [bs4](https://www.crummy.com/software/BeautifulSoup/), which can recover from some syntax errors but they alter the resulting XML in the process. Also, these parsers try to get a typical DOM out of the XML file and then use it to search for a particular element or set of elements. Here we are more interested in getting all the metadata for each kind of token (like the exact position in the document) so we are more interested in getting a syntax tree than a regular XML DOM.

I've been researching existing language servers that deal with similar parsing techniques like [Microsoft VSCode HTML language service](https://github.com/microsoft/vscode-html-languageservice) or the [Eclipse XML language server, Lemminx, implemented in Java](https://github.com/eclipse/lemminx) so I will try to port this kind of solution to Python and see how it works.

### PR DESCRIPTION
**Pull request by davelopez**
This (huge?) PR replaces the old SAX parser with a more stable implementation based on the JAVA [Eclipse/Lemminx](https://github.com/eclipse/lemminx/tree/master/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/parser) parser implementation (which is also based in the [Microsoft's HTML parser](https://github.com/microsoft/vscode-html-languageservice/tree/master/src/parser)).

This implementation provides much more detail about the position of every token, even if the XML is incomplete or malformed.
Also, some related classes, like `XmlContext` were updated and improved to benefit from the new parser.

Finally, some new tests were added and most of the ones that were still relevant adapted to cover a wide range of cases.

Closes #53

### CODE REVIEW

File: server/galaxyls/services/xml/utils.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                return True
+            self.advance(1)
+        return False
+
+    def advance_until_chars(self, ch: List[int]) -> bool:
+        """Advances the stream until it founds a list of character matching the given."""
+        while self._position + len(ch) <= self._len:
+            i = 0
+            while i < len(ch) and self._source[self._position + i] == chr(ch[i]):
+                i = i + 1
```
Comment (bgruening) [inline server/galaxyls/services/xml/utils.py:79]: i += 1

File: server/galaxyls/services/xml/utils.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            if i == len(ch):
```
Comment (bgruening) [inline server/galaxyls/services/xml/utils.py:80]: You could take the len() calculation out of the while look

File: server/galaxyls/services/xml/utils.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                return True
+            self.advance(i or 1)
+        self.go_to_end()
+        return False
+
+    def advance_while_char_in(self, list: List[int]) -> int:
+        """Advances the stream if the characters are like any of the characters in the given list."""
+        pos_now = self._position
+        while self._position < self._len and ord(self._source[self._position]) in list:
+            self._position = self._position + 1
```
Comment (bgruening) [inline server/galaxyls/services/xml/utils.py:90]: +=

File: server/galaxyls/services/xml/utils.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    def advance_until_char_or_new_tag(self, ch: int) -> bool:
+        """Advances the stream until it finds the given character or the '<' (new tag character)."""
+        while self._position < self._len:
+            if self.peek_char() == ch or self.peek_char() == _LAN:
+                return True
+            self.advance(1)
+        return False
+
+    def advance_until_chars_or_new_tag(self, ch: List[int]) -> bool:
```
Comment (bgruening) [inline server/galaxyls/services/xml/utils.py:101]: Some redundancy with advance_until_chars?

File: server/galaxyls/services/xml/utils.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                return True
+            self.advance(i or 1)
+        self.go_to_end()
+        return False
+
+    def advance_while_char(self, predicate: Callable[[str], bool]) -> int:
+        """Advances the stream while the given condition is True."""
+        pos_now = self._position
+        while self._position < self._len and predicate(self._source[self._position]):
+            self._position = self._position + 1
```
Comment (bgruening) [inline server/galaxyls/services/xml/utils.py:119]: +=

File: server/galaxyls/services/xml/utils.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        """Gets the source text."""
+        return self._source
+
+    def pos(self) -> int:
+        """Gets the current stream position."""
+        return self._position
+
+    def advance(self, n: int) -> None:
+        """Advances the stream given number of positions."""
+        self._position = self._position + n
```
Comment (bgruening) [inline server/galaxyls/services/xml/utils.py:34]: self._position += n

File: server/galaxyls/services/xml/parser.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            token = scanner.scan()
+
+        if previous_token_was_end_tag_open:
+            previous_token_was_end_tag_open = False
+            if token != TokenType.EndTag:
+                # The expected token is not an EndTag, create a fake end tag element
+                self._create_fake_end_tag(end_tag_open_offset, current)
+
+        while current.parent:
+            current.end = len(text)
```
Comment (bgruening) [inline server/galaxyls/services/xml/parser.py:219]: len does not change inside the while, does it?

**Comment by bgruening** [comment]
:heart:

**Comment by davelopez** [comment]
Thanks for taking the time to review @bgruening!
I may have translated these algorithms from JAVA to Python too literally ;)

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Awesome! This will give us some more freedom! :)

**Pull request merged**